### PR TITLE
PIM-5753: Array converters ordered by tag priority

### PR DESCRIPTION
**Pull request by jmleroux**
**Description**

This PR add an ordering feature for arrays converters.
Related issue: PIM-5753 Accept numeric values for text attributes in import

Btw, we should add position tags to all of our registries to allow external extension to register their services in the right place.

**Definition Of Done (for Core Developer only)**
_As a traitor, i will fill the DoD anyway_

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       |only modified
| Added Behats                      |no
| Changelog updated                 |yes
| Review and 2 GTM                  |
| Micro Demo to the PO (Story only) |no
| Migration script                  |no
| Tech Doc                          |no

### CODE REVIEW

File: src/Pim/Bundle/ConnectorBundle/DependencyInjection/Compiler/RegisterConverterPass.php [hunk 1]
```diff
@@ -4,10 +4,11 @@
 
 use Symfony\Component\DependencyInjection\Compiler\CompilerPassInterface;
 use Symfony\Component\DependencyInjection\ContainerBuilder;
+use Symfony\Component\DependencyInjection\Definition;
 use Symfony\Component\DependencyInjection\Reference;
 
 /**
- * Register converters
+ * Register converters ordered by position
```
Comment (aRn0D) [inline src/Pim/Bundle/ConnectorBundle/DependencyInjection/Compiler/RegisterConverterPass.php]: Where is the spec?

File: src/Pim/Bundle/ConnectorBundle/DependencyInjection/Compiler/RegisterConverterPass.php [hunk 2]
```diff
@@ -34,11 +38,31 @@ public function process(ContainerBuilder $container)
...
+
+    /**
+     * @param Definition $registry
+     * @param string     $serviceId
+     * @param string[]   $tags
+     */
+    protected function registerConverter($registry, $serviceId, $tags)
+    {
+        foreach ($tags as $tag) {
+            $position = isset($tag['position']) ? (int)$tag['position'] : static::DEFAULT_POSITION;
```
Comment (aRn0D) [inline src/Pim/Bundle/ConnectorBundle/DependencyInjection/Compiler/RegisterConverterPass.php]: self instead static ?

File: src/Pim/Bundle/ConnectorBundle/DependencyInjection/Compiler/RegisterConverterPass.php [hunk 3]
```diff
@@ -34,11 +38,31 @@ public function process(ContainerBuilder $container)
...
+            $this->registerConverter($registry, $serviceId, $tags);
+        }
+    }
+
+    /**
+     * @param Definition $registry
+     * @param string     $serviceId
+     * @param string[]   $tags
+     */
+    protected function registerConverter($registry, $serviceId, $tags)
```
Comment (aRn0D) [inline src/Pim/Bundle/ConnectorBundle/DependencyInjection/Compiler/RegisterConverterPass.php]: Definition $registry instead $registry ?

File: src/Pim/Bundle/ConnectorBundle/DependencyInjection/Compiler/RegisterConverterPass.php (hunk 1 above)
Comment (jmleroux) [inline src/Pim/Bundle/ConnectorBundle/DependencyInjection/Compiler/RegisterConverterPass.php]: @aRn0D , in master branch i suppose. 

Ok, my bad, i'll also backport it. Good catch :+1:

Comment (juliensnz) [inline src/Pim/Bundle/ConnectorBundle/DependencyInjection/Compiler/RegisterConverterPass.php]: as said earlier, position make sens for visual stuff but should be priority IMHO

File: src/Pim/Component/Connector/ArrayConverter/Flat/Product/ValueConverter/ValueConverterRegistryInterface.php [hunk 4]
```diff
@@ -15,10 +17,11 @@
      * Register a converter.
      *
      * @param ValueConverterInterface $converter
+     * @param int                     $position FIFO order : converter with lower priority is inspected first.
      *
      * @return ValueConverterRegistry
      */
-    public function register(ValueConverterInterface $converter);
+    public function register(ValueConverterInterface $converter, $position = RegisterConverterPass::DEFAULT_POSITION);
```
Comment (aRn0D) [inline src/Pim/Component/Connector/ArrayConverter/Flat/Product/ValueConverter/ValueConverterRegistryInterface.php]: you are using a constant from a compiler pass in a component!

File: src/Pim/Bundle/ConnectorBundle/DependencyInjection/Compiler/RegisterConverterPass.php (hunk 2 above)
Comment (juliensnz) [inline src/Pim/Bundle/ConnectorBundle/DependencyInjection/Compiler/RegisterConverterPass.php]: space after (int)

Comment (jmleroux) [inline src/Pim/Bundle/ConnectorBundle/DependencyInjection/Compiler/RegisterConverterPass.php]: @aRn0D , i tend to prefer `static` but to be honest we don't care: this class is not meant to be overriden. It should even be private (/cc @jjanvier)

Comment (jmleroux) [inline src/Pim/Bundle/ConnectorBundle/DependencyInjection/Compiler/RegisterConverterPass.php]: @juliensnz , yes but no: Travis is OK without the sapec, so is my Storm auto format. So be it.

File: CHANGELOG-1.4.md [hunk 5]
```diff
@@ -1,3 +1,11 @@
+# 1.4.x
+
+## Improvements
+- PIM-5753: Add a `position` tag parameter for `ValueConverterRegistry` to allow ordering of array converters.
```
Comment (juliensnz) [inline CHANGELOG-1.4.md]: you changed to priority ?

**Comment by juliensnz** [comment]
GTM

File: CHANGELOG-1.4.md (hunk 5 above)
Comment (jmleroux) [inline CHANGELOG-1.4.md]: yes ;)

**Comment by nidup** [comment]
'position' to 'priority' everywhere then GTM.

about generic registries we talked about this yesterday with @juliensnz and @aRn0D looks we'll provide few generic one in a new component (or using an existing component from Sylius for instance).

**Pull request merged**